refactor: use WidgetsIcon for null resources

jsjoeio opened the pull request:
This PR updates the icon used for null resources 

<img width="59" alt="image" src="https://user-images.githubusercontent.com/3806031/187471925-7c3d73a9-2876-4fca-a770-89b5c2241581.png">

Previous icon used:

<img width="49" alt="image" src="https://user-images.githubusercontent.com/3806031/187471858-ee56ce89-8848-404e-b23e-4b9719931b9f.png">


h/t to @kconley-sq for the suggestion! 🎉

Also found another issue which I will tackle after this: https://github.com/coder/coder/issues/3767

bpmct commented:
NICE! 👍🏼

bpmct commented:
@jsjoeio is the unknown resource still a `?` or does this also apply to unknown resources? (e.g. aws_spot_instance_request)

<img width="1344" alt="Screen Shot 2022-08-30 at 10 13 17 AM" src="https://user-images.githubusercontent.com/22407953/187474401-f0dbbdab-4dc8-44fe-bd00-31b94b8b4759.png">

bpmct commented:
resolves #3301

jsjoeio commented:
> Edit: answered my own question, I think we should changed this to the widget icon too

Nice catch! I'll update that now then re-request a review.

jsjoeio commented on site/src/components/Resources/ResourceAvatar.tsx:
```diff
@@ -22,15 +24,13 @@ const iconByResource: Record<string, typeof MemoryIcon> = {
-  // this resource can return undefined
-  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-  const IconComponent = iconByResource[type] ?? HelpIcon
+  const IconComponent = iconByResource[type] ?? WidgetsIcon
```
Since the resource can be `undefined`, I modified the return type of `iconByResource` to match that, which lets us remove this `eslint` line.

jsjoeio commented on site/src/components/Resources/ResourceAvatar.tsx:
```diff
@@ -22,15 +24,13 @@ const iconByResource: Record<string, typeof MemoryIcon> = {
 }
 
-export type ResourceAvatarProps = { type: string }
+export type ResourceAvatarProps = { type: ValidIcons }
```
`string` is too wide. By creating `ValidIcons`, we have a string union that accurately reflects `type`.

bpmct replied:
looks like a test is now failing?

jsjoeio replied:
> Type 'string' is not assignable to type 'ValidIcons'.

Ahh...I think the type in the component props needs to be updated.

jsjoeio replied:
Okay, so it turns out there is a mismatch between types generated by backend and expectations on frontend. Backend says it can be of type `string` but really, it is a limited set of strings for which we have Icons.

Moving to draft so I can fix that first.

jsjoeio replied:
I decided that will take more work and should be done as a separate PR so I'll clean this up and open a separate issue to address the types mismatch.

jsjoeio replied:
https://github.com/coder/coder/issues/3767

jsjoeio committed:
refactor: replace HelpIcon w/WidgetsIcon

Based on user feedback, we believe the `WidgetsIcon` will cause less
confusion.

jsjoeio committed:
fixup

jsjoeio committed:
refactor: clean up types in ResourceAvatar.tsx

Before, we were using `string` for `type` in `ResourceAvatar`. This
meant it wasn't tied to the types generated from the backend.

Now it imports `WorkspaceResource` so that there is a single source of
truth and they always stay in sync.
